### PR TITLE
Part: Persistent section analysis view

### LINKED ISSUE #5570: Parts Appear as Shells Instead of Solids When Viewed by Clipping Plane
Issue imported from https://tracker.freecad.org/view.php?id=1208

* **Reporter:** bejant
* **Date submitted:** 8/10/2013
* **FreeCAD version:** 0.13
* **Category:** Feature
* **Status:** new
* **Tags:** 

# Original report text

FreeCAD version is actually 0.14 (0.14 presently is not selectable from Product Version box).

I do not know if the parts are actually solids that appear as shells, of if the parts really are hollow shells.  Link to forum thread:

http://forum.freecadweb.org/viewtopic.php?f=3&t=4371

# Other bug information

* **Priority:** normal
* **Severity:** minor
* **Category:** Feature
* **Updated:** 5/9/2021

# Discussion from Mantis ticket

### Comment by Kunda1 2017-01-13 13:29
Some developments with coin3d has made this a possiblity: Please see http://forum.freecadweb.org/viewtopic.php?f=3&t=4371&start=10#p94838

---

### Comment by berndhahnebach 2019-03-01 21:32
related https://forum.freecadweb.org/viewtopic.php?f=8&t=18188

---

### Comment by dstarke 2021-05-09 20:42
Create a rectangular filled face in the Draft workbench and use that with the "Slice apart" tool of the Part workbench also results in a solid clipping plane.
See https://www.xsim.info/articles/FreeCAD/en-US/HowTo/Cut-a-solid-with-a-plane.html

### PR DESCRIPTION
**Pull request by greggjaskiewicz**
## What it does

Fusion-style section analysis for FreeCAD. Clips a shape at a cutting plane, renders filled cross-section faces with diagonal hatching. Lives in the feature tree, available from the View menu in any workbench. Persistent.

## How it works 

OCCT boolean cut (half-space subtraction) computes filled cross-section faces at the cutting plane — those become the hatched/colored geometry. Separately, a Coin3D SoClipPlane clips the source object's rendering at the same plane so you see inside.

## How to use
Select the part/tree you want to create section view of.  Go to view->section analysis - and follow directions on screen. 

## Benefits
Persistence, availability across WB, etc. 

## Issues
It's not as fast as rendering tricks used currently. 
There are some issues with gizmos adjusting angles. I'm not familiar with those enough to address them, certainly welcome feedback. 

Potentially fixes https://github.com/FreeCAD/FreeCAD/issues/5570

Screenshot 2026-03-25 at 18 52 08

Screenshot 2026-03-25 at 18 59 01

Screenshot 2026-03-25 at 18 22 04

Screenshot_2026-03-20_at_22 25 16

### CODE REVIEW

**Comment by hyarion** [comment]
> There was a question about it being similar Persistent section cut in Parts WB. Unfortunately that tool never worked well for me, and this basically replaces it - but better.

Persistent section cut is meant to cut the body so you can export it.
Your tool is for visual analysis. I'd just call it Section analysis instead :)

**Comment by mno13** [comment]
Is there an option for select face/plane or axis as reference plane, and just modify offset from selection? 

I see only preset as drop-down, and manual way to enter normals. From my experience - when section is not done on XYZ planes,  it almost always done as paralel to existing face or thru shaft/hole axis.

**Comment by pierreporte** [comment]
I confirm that this PR is subject to #22542.

**Comment by pierreporte** [comment]
@greggjaskiewicz have a look at #29294. It could be an alternative to have hatches filling the shape in 2D like in Fusion 360. Though I would prefer the Fusion 360 way for section view.

**Comment by PaddleStroke** [comment]
What is the path forward for this? Is there still issues to be fixed? This feature has been asked many time in the past

**Comment by belinski-roman** [comment]
When should we wait for the merger?

**Comment by greggjaskiewicz** [comment]
the feature needs review from freeCAD devs and any suggestions they might have. But I wouldn't be offended if they took it and merged it as is - and worked on any follow up fixes in main

**Comment by hyarion** [comment]
> the feature needs review from freeCAD devs and any suggestions they might have. But I wouldn't be offended if they took it and merged it as is - and worked on any follow up fixes in main

Oh, I didn't know it was ready for review

**Comment by greggjaskiewicz** [comment]
> Oh, I didn't know it was ready for review

Well, I need some feedback to make changes necessary to get this in. I of course will have to rebase this - the is way too many commits - and some of it was just nonsense changes or such

**Comment by pierreporte** [comment]
There are a lot of changed that I suggested. Precise positioning of the cutting plane is major.

**Comment by deltahedra3d** [comment]
I’m currently testing the feature and I think it’s great! However, I noticed some strange behavior with Part Design bodies containing multiple solids: the cross-hatching doesn't appear on all of them. Also, when checking/unchecking boxes, other hatch patterns suddenly appear.
section

**Review by Prazgreen1 — APPROVED** [review]

**Comment by Prazgreen1** [comment]
Sorry, my bad. I misclicked on that.

**Comment by pierreporte** [comment]
I just found this OCCT documentation page where it explictly mentions clipping plane that are filled with some sort of image. Maybe it already has a lot of things needed, that the PR could reuse?

https://dev.opencascade.org/doc/overview/html/index.html#intro_overview_visu

Ping @dpasukhi

**Comment by dpasukhi** [comment]
Yes, OCCT provided that option. But it is not some neutral feature, it requires usage of OCCT renderer and build around it. As I know you don't use it deeply and implemention would be significantly different (even probably shader)
It use some internal contexts of view (including all objects) and their information, it is very complex cooperation of things :(

**Comment by greggjaskiewicz** [comment]
That would be awesome . I haven’t seen it, but it’s a massive library .I will be back at this towards end of the month . Currently away on holidays  On 5 Jun 2026, at 23:40, Thomas D. ***@***.***> wrote:﻿pierreporte left a comment (FreeCAD/FreeCAD#28647)
I just found this OCCT documentation page where it explictly mentions clipping plane that are filled with some sort of image. Maybe it already has a lot of things needed, that the PR could reuse?
https://dev.opencascade.org/doc/overview/html/index.html#intro_overview_visu

**Comment by iuvi7** [comment]
Great work ! 🔥 Especially 1.2 on the way of rolling next

**Comment by pierreporte** [comment]
@greggjaskiewicz could you give me your opinion on my comments in https://github.com/FreeCAD/FreeCAD/pull/28647#issuecomment-4137265148?

**Comment by greggjaskiewicz** [comment]
Just some responses, 
Agreed — OCCT capping is tied to OCCT's OpenGl viewer and full-scene context, which FreeCAD's Coin3D pipeline doesn't use, so reusing it would mean either
  adopting the OCCT renderer or writing custom Coin3D shaders.
 The boolean-cut + SoClipPlane approach in this PR is the renderer-compatible way to get filled sections today.

The multiple body issue is something I am going to test and see if I can fix . Standby !

**Comment by greggjaskiewicz** [comment]
pushed a small update, mainly fixing the per-body distinction that has been annoying me to no end. I am going to spend some time going over the whole change again - since I forgot a lot about what was happening in this code, and squash it into single commit. There have been loads of changes and so many commits . Some of them are  embarrassing. :)

**Comment by codeotaku** [comment]
I've also been testing - this is great - really nice work!

One thing I noticed is that the textures for the hash lines appear quite pixelated when zooming in.  With some experimentation, I was able to use vectors _SoIndexedLineSet_ instead of the _SoTexture2_ generate the hash lines, so that they are crisp and uniform at every zoom level (as would be expected for vectors).  

Screens: 


**Texture** with  _SoTexture2_```const int sz = 16``` (line 173) gives a blurry appearance:

Screenshot 2026-08-06 at 3 39 55 PM


**Vector** using _SoIndexedLineSet_ gives (at two zoom levels):

Screenshot 2026-08-06 at 4 22 43 PM

Screenshot 2026-08-06 at 4 22 37 PM

Again, really great work on this.  

I'll share the reference code.

**Comment by greggjaskiewicz** [comment]
thanks @codeotaku , I'll have a look. Yes, it is only thing that annoys me - the blurriness. I can see it on the mac, but its not as bad on linux . I'm sure its just a monitor quality, dell vs apple.

**Comment by greggjaskiewicz** [comment]
do I need to add any tests? It's weird writing code in not so TDD style here? 
Do I need to squash everything into one commit ? 
Any other requirements before this could be accepted?

**Comment by greggjaskiewicz** [comment]
I might push some further updates to this, might have found a way to make it even smaller

**Comment by maxwxyz** [comment]
Thanks, this is already a substantial improvement. I compared it with [#16187](https://github.com/FreeCAD/FreeCAD/issues/16187) and its sub-issues, which should define the overall scope. I see this PR primarily as an implementation of #16187, rather than a solution to the closed-profile issue currently linked.

Persistent section objects, multiple simultaneous sections, the visible plane, flip direction, per-body treatment, and a usable section result are all valuable. I definitely want to preserve this work.
I would nevertheless prefer a small architectural change. This should become the first implementation of the persistent clipping-plane concept from #16187, rather than a Part-only SectionAnalysis object that combines generic clipping state, the Coin clipping implementation, and a Part shape result.
At present, the object owns Source, PlaneNormal, PlaneOffset, FlipCut, the generated section faces, and the rendering behavior. That works for the current use case, but makes later work on saved views, attachments, target filtering, and reusable clipping planes unnecessarily difficult.

Could the design be split into:

- A persistent document-level clipping-plane object, with an enabled state, placement, reverse direction, a visible plane, and multiple active instances
- A target policy that defaults to the whole document and can later support explicit objects to cut and objects to exclude, as described in #16190
- A GUI view provider that installs and removes the Coin clip plane and supports editing by double-click and direct manipulation
- An optional Part section/profile result that references the clipping plane and produces capped, hatched geometry only when requested, rather than making the expensive boolean result the defining feature of every clipping plane

In particular, the interactive clipping and visual closed face should be rendered in Coin. Dragging the plane through an assembly should update only the visual clipping plane, not repeatedly calculate a Part section. This would remove the current lag while dragging and would also allow visual clipping and capping for non-solid objects, such as meshes.

The optional parametric profile should be generated only when the user explicitly requests it, for example with a Generate profile outline option when accepting the task panel. It would then remain useful for measurements, export, sketches, and downstream features, as requested in #16193.

Using a Placement as the primary plane definition would also make the object easier to attach to faces, datum planes, axes later, addressing #11169 and #11173. PlaneNormal and PlaneOffset can remain convenient UI representations, but should not be the only persistent model.

It is fine if camera and visibility snapshots, attachment UI, profile export, additional closed-cap options, and finite clipping extents arrive in follow-up PRs. However, the data model should leave room for them now. A stored view snapshot should eventually restore camera and object visibility, while the section profile must remain parametric and update from the same clipping-plane object.
One detail to keep separate: the displayed plane size is a visual aid, while a finite X/Y clipping extent is a different feature from an infinite Coin clipping plane and should not be accidentally mixed with it.

**Comment by greggjaskiewicz** [comment]
@maxwxyz what would be the benefit of splitting it up? It's all-or-nothing feature, no point having 2/3rds of it in for instance. 
I did simplify it, for sake of being ready to merge - despite there being a lot of requests for various little features - which could easily be done afterwards.
Even tho I'm a seasoned software dev, this is first time I'm trying to get something this big into FreeCAD - so I appreciate that I might not know how things work around here.
 
Re tests, I got some tests I wrote, and some that were vibe-added - and things got fixed that way that would be hard to find.

**Comment by maxwxyz** [comment]
@greggjaskiewicz no I did not meant to split up the current PR but maybe the design of the tool to accommodate for the later additions.

**Comment by greggjaskiewicz** [comment]
@maxwxyz of course, I dont think it will be hard to expand on it. I'm happy to split things up further, but some things have to be in tight loops for the performance sake. 
Should I do a squash? This PR has evolved so many times, I feel like old versions are irrelevant

**Comment by maxwxyz** [comment]
Yes sure you can squash or, if possible, make meaningful commits.

**Comment by greggjaskiewicz** [comment]
@maxwxyz agreed on Placement, we should store as little as possible in the document - and make it as future proof as possible. This is one of those things, I got no experience as far as FreeCAD is concerned . So I'll try something, but I'm open to changes or suggestions.

**Comment by greggjaskiewicz** [comment]
force pushed with all changes I wanted to make, and it should now just be couple of separate discrete commits. Doing this during my lunch break, so hopefully I Didnt break anything :D Oh, and sorry about the number of tests - but I kinda went crazy. A lot of them were suggested by claude - since predicting edge cases isnt always nice. But testing this manually each time is a pain.

**Comment by greggjaskiewicz** [comment]
I believe this also answers a lot of @pierreporte 's feedback.

File: src/Mod/Part/Gui/Command.cpp [hunk 1]
```diff
@@ -2446,6 +2452,154 @@ bool CmdPartProjectionOnSurface::isActive()
...
+
+    std::string docName = getDocument()->getName();
+    std::string sourceList;
+    for (auto* obj : sources) {
+        if (!sourceList.empty()) {
+            sourceList += ", ";
+        }
+        sourceList += "App.getDocument('" + docName + "').getObject('" + obj->getNameInDocument()
+            + "')";
+    }
```
Comment (kadet1090) [inline src/Mod/Part/Gui/Command.cpp:2539]: @chennes is this safe? i.e. can it be used to do arbitrary code execution?

File: src/Mod/Part/Gui/TaskSectionAnalysis.cpp [hunk 2]
```diff
@@ -0,0 +1,653 @@
...
+    presetCombo->addItem(tr("XY Plane (Z normal)"));
+    presetCombo->addItem(tr("XZ Plane (Y normal)"));
+    presetCombo->addItem(tr("YZ Plane (X normal)"));
+    presetCombo->addItem(tr("View Direction"));
+    presetCombo->addItem(tr("Custom"));
+    planeLayout->addWidget(presetCombo, 0, 1);
+
+    // Detect current preset from normal
+    Base::Vector3d n = feature->PlaneNormal.getValue();
+    if (std::abs(n.x) < 1e-6 && std::abs(n.y) < 1e-6 && std::abs(std::abs(n.z) - 1.0) < 1e-6) {
```
Comment (kadet1090) [inline src/Mod/Part/Gui/TaskSectionAnalysis.cpp]: 1e-6 can be probably replaced with something like `Precision::confusion()`

File: src/Mod/Part/Gui/TaskSectionAnalysis.cpp [hunk 3]
```diff
@@ -0,0 +1,653 @@
...
+    }
+    else if (std::abs(n.x) < 1e-6 && std::abs(std::abs(n.y) - 1.0) < 1e-6 && std::abs(n.z) < 1e-6) {
+        presetCombo->setCurrentIndex(1);
+    }
+    else if (std::abs(std::abs(n.x) - 1.0) < 1e-6 && std::abs(n.y) < 1e-6 && std::abs(n.z) < 1e-6) {
+        presetCombo->setCurrentIndex(2);
+    }
+    else {
+        presetCombo->setCurrentIndex(4);  // Custom
+    }
```
Comment (kadet1090) [inline src/Mod/Part/Gui/TaskSectionAnalysis.cpp]: I'd love to see it in a bit more readable style. These conditions are not self explanatory and I have no idea what they mean especially that using raw indices is also not self-explanatory. I see two improvements here to be made:

1. Introduce `enum` (not `enum class`) for the preset combo so you can use well named cases
2. Write a comment on top that explain what's the reasoning between all of them - or wrap them in some local lambdas / variables with self-explanatory names.

File: src/Mod/Part/Gui/TaskSectionAnalysis.cpp [hunk 4]
```diff
@@ -0,0 +1,653 @@
...
+    return viewProvider;
+}
+
+
+Part::SectionAnalysis* SectionAnalysisWidget::getObject() const
+{
+    return feature;
+}
+
+void SectionAnalysisWidget::setupUi()
```
Comment (kadet1090) [inline src/Mod/Part/Gui/TaskSectionAnalysis.cpp:160]: Would it be possible to use .ui file here? While I see benefits of doing it in code and it's way easier to read for me - in general we use the .ui files and should stick with it.

File: src/Mod/Part/Gui/TaskSectionAnalysis.cpp [hunk 5]
```diff
@@ -0,0 +1,653 @@
...
+            return;
+    }
+
+    // Enable angle spinboxes for axis-aligned presets
+    angle1Spin->setEnabled(true);
+    angle2Spin->setEnabled(true);
+
+    // Update angle labels for the selected preset
+    switch (index) {
+        case 0:  // XY (Z normal)
```
Comment (kadet1090) [inline src/Mod/Part/Gui/TaskSectionAnalysis.cpp]: enum would be better than comments

File: src/Mod/Part/Gui/TaskSectionAnalysis.cpp [hunk 6]
```diff
@@ -0,0 +1,653 @@
...
+    connect(hatchCheck, &QCheckBox::toggled, this, &SectionAnalysisWidget::onHatchToggled);
+    connect(autoHideHatchCheck, &QCheckBox::toggled, this, [this](bool on) {
+        viewProvider->AutoHideHatching.setValue(on);
+    });
+    connect(perSolidColorCheck, &QCheckBox::toggled, this, &SectionAnalysisWidget::onPerSolidColorToggled);
+    connect(showPlaneCheck, &QCheckBox::toggled, this, &SectionAnalysisWidget::onShowPlaneToggled);
+    connect(updateViewCheck, &QCheckBox::toggled, this, &SectionAnalysisWidget::onUpdateViewToggled);
+}
+
+void SectionAnalysisWidget::onPresetChanged(int index)
```
Comment (kadet1090) [inline src/Mod/Part/Gui/TaskSectionAnalysis.cpp:593]: This function does quite a lot, I'd split it into separate private methods per concern. Then it would be way easier for you to use RAII `QSignalBlocker` instead of manually changing the flag.

File: src/Mod/Part/Gui/TaskSectionAnalysis.cpp [hunk 7]
```diff
@@ -0,0 +1,653 @@
...
+    switch (index) {
+        case 0:
+            normal = Base::Vector3d(0, 0, 1);
+            break;
+        case 1:
+            normal = Base::Vector3d(0, 1, 0);
+            break;
+        case 2:
+            normal = Base::Vector3d(1, 0, 0);
+            break;
```
Comment (kadet1090) [inline src/Mod/Part/Gui/TaskSectionAnalysis.cpp]: Here you can use `Vector3d::UnitX`, `Vector3D::UnitY` etc for better readability. Also enums

File: src/Mod/Part/Gui/TaskSectionAnalysis.cpp [hunk 8]
```diff
@@ -0,0 +1,653 @@
...
+        case 3: {
+            // View Direction: get camera direction
+            auto* mdiView = qobject_cast<Gui::View3DInventor*>(
+                Gui::Application::Instance->activeView()
+            );
+            if (mdiView) {
+                SbVec3f vd = mdiView->getViewer()->getViewDirection();
+                float vx, vy, vz;
+                vd.getValue(vx, vy, vz);
+                normal = Base::Vector3d(-vx, -vy, -vz);  // face toward camera
```
Comment (kadet1090) [inline src/Mod/Part/Gui/TaskSectionAnalysis.cpp]:
```suggestion
                normal = -Base::convertTo<Base::Vector3d>(mdiView->getViewer()->getViewDirection()))
```

That also should work?

File: src/Mod/Part/Gui/TaskSectionAnalysis.cpp [hunk 9]
```diff
@@ -0,0 +1,653 @@
...
+    recompute();
+}
+
+void SectionAnalysisWidget::onSectionColorChanged(const QColor& color)
+{
+    // Update the section face color on the ViewProvider
+    auto* vp = dynamic_cast<PartGui::ViewProviderSectionAnalysis*>(
+        Gui::Application::Instance->getViewProvider(feature)
+    );
+    if (vp) {
```
Comment (kadet1090) [inline src/Mod/Part/Gui/TaskSectionAnalysis.cpp]: I think that `getViewProvider` has templated instance so you could do:


```suggestion
    if (auto* vp = Gui::Application::Instance->getViewProvider<PartGui::ViewProviderSectionAnalysis>(feature)) {
```

I also suggest using if init-statements so it's impossible to use the unchecked vp

File: src/Mod/Part/Gui/TaskSectionAnalysis.cpp [hunk 10]
```diff
@@ -0,0 +1,653 @@
...
+void SectionAnalysisWidget::updateFromFeature()
+{
+    // Sync controls from the feature without triggering signals; decomposes
+    // the normal into angles relative to the current preset.
+    Base::Vector3d n = feature->PlaneNormal.getValue();
+    bool flip = feature->FlipCut.getValue();
+
+    flipCheck->blockSignals(true);
+    angle1Spin->blockSignals(true);
+    angle2Spin->blockSignals(true);
```
Comment (kadet1090) [inline src/Mod/Part/Gui/TaskSectionAnalysis.cpp]: QSignalBlocker would be preferred here

File: src/Mod/Part/Gui/TaskSectionAnalysis.cpp [hunk 11]
```diff
@@ -0,0 +1,653 @@
...
+    int preset = presetCombo->currentIndex();
+    switch (preset) {
+        case 0: {  // XY plane (Z normal)
+            // Forward: n = (cosA*sinB, sinA, cosA*cosB)
+            Base::Vector3d ne = (n.z < 0) ? -n : n;
+            double alpha = std::asin(std::clamp(ne.y, -1.0, 1.0));
+            double cosA = std::cos(alpha);
+            double beta = (cosA > 1e-10) ? std::atan2(ne.x, ne.z) : 0.0;
+            angle1Spin->setValue(alpha * 180.0 / M_PI);
+            angle2Spin->setValue(beta * 180.0 / M_PI);
```
Comment (kadet1090) [inline src/Mod/Part/Gui/TaskSectionAnalysis.cpp]: `std::numbers::pi` should be used - I recommend doing `using` statement though.

File: src/Mod/Part/Gui/ViewProviderSectionAnalysis.cpp [hunk 12]
```diff
@@ -0,0 +1,1245 @@
...
+    pcPlaneRoot->addChild(pickStyle);
+
+    pcPlaneHints = new SoShapeHints();
+    pcPlaneHints->vertexOrdering = SoShapeHints::COUNTERCLOCKWISE;
+    pcPlaneHints->shapeType = SoShapeHints::UNKNOWN_SHAPE_TYPE;
+    pcPlaneRoot->addChild(pcPlaneHints);
+
+    pcPlaneMaterial = new SoMaterial();
+    pcPlaneMaterial->diffuseColor.setValue(0.3f, 0.6f, 0.9f);
+    pcPlaneMaterial->transparency.setValue(0.7f);
```
Comment (kadet1090) [inline src/Mod/Part/Gui/ViewProviderSectionAnalysis.cpp]: I wonder if this should be themeable?

File: src/Mod/Part/Gui/ViewProviderSectionAnalysis.cpp [hunk 13]
```diff
@@ -0,0 +1,1245 @@
...
+                static_cast<float>(localNormal.x),
+                static_cast<float>(localNormal.y),
+                static_cast<float>(localNormal.z)
+            ),
+            SbVec3f(
+                static_cast<float>(localPoint.x),
+                static_cast<float>(localPoint.y),
+                static_cast<float>(localPoint.z)
+            )
+        ));
```
Comment (kadet1090) [inline src/Mod/Part/Gui/ViewProviderSectionAnalysis.cpp]: You can use `Base::convertTo` here

File: src/Mod/Part/Gui/ViewProviderSectionAnalysis.cpp [hunk 14]
```diff
@@ -0,0 +1,1245 @@
...
+    auto isAncestorOrSelf = [](const App::DocumentObject* a, const App::DocumentObject* b) {
+        for (const auto* o = b; o; o = App::GeoFeatureGroupExtension::getGroupOfObject(o)) {
+            if (o == a) {
+                return true;
+            }
+        }
+        return false;
+    };
+    const auto& sources = feat->Source.getValues();
+    const bool related = std::any_of(sources.begin(), sources.end(), [&](const auto* src) {
```
Comment (kadet1090) [inline src/Mod/Part/Gui/ViewProviderSectionAnalysis.cpp:1151]: std::ranges can be used here for a bit better readability

File: src/Mod/Part/Gui/ViewProviderSectionAnalysis.cpp [hunk 15]
```diff
@@ -0,0 +1,1245 @@
...
+    Bnd_Box bbox;
+    if (!feat->sourceBoundingBox(bbox)) {
+        return;
+    }
+
+    bbox.Get(sourceBBox[0], sourceBBox[1], sourceBBox[2], sourceBBox[3], sourceBBox[4], sourceBBox[5]);
+    sourceBBoxValid = true;
+}
+
+void ViewProviderSectionAnalysis::updatePlaneVisual()
```
Comment (kadet1090) [inline src/Mod/Part/Gui/ViewProviderSectionAnalysis.cpp:1259]: I feel that this is a bit repetitive with what I did for Show Plane here 

https://github.com/FreeCAD/FreeCAD/blob/a2468774d39b52d41714e641fc607ed1773d428d/src/Mod/Part/Gui/ViewProvider2DObject.cpp#L440-L521 

I wonder if there is a way to make something more reausable.

File: src/Mod/Part/Gui/ViewProviderSectionAnalysis.cpp [hunk 16]
```diff
@@ -0,0 +1,1245 @@
...
+    transformDragger->setUpAutoScale(viewer->getSoRenderManager()->getCamera());
+
+    // Orient the dragger so Z points out of the cut, away from the material.
+    // Translation along Z = change PlaneOffset.
+    // Rotation around X/Y = change PlaneNormal.
+    // This was tricky to get right
+    Base::Vector3d n;
+    double d = 0.0;
+    if (getCutFrame(n, d)) {
+        Base::Rotation rot(Base::Vector3d(0, 0, 1), n);
```
Comment (kadet1090) [inline src/Mod/Part/Gui/ViewProviderSectionAnalysis.cpp]: `Base::Vector3d::UnitZ`

**Review by kadet1090 — CHANGES_REQUESTED** [review]
First round of review - some comments, I'd say that mostly minor and nitpicks. The biggest issue I see is that a lot of geometry preparation seems to be duplicated - I wonder if there is a better way to do it?

Apart from that - it's really awesome PR to read, very easy to follow and understand. I'd love to see more work like that in FreeCAD!

**Comment by pierreporte** [comment]
Just compiled. I have the following issues:

Using the assembly example file, solids created at the root of the document have the filling but not the assembly container.

image


Hatches seem to be slightly above the cutting plane.

image


There shouldn’t be an error if there is no intersection with geometry. Just let the cutting plane sit in the empty.


The gizmo has an empty Z-axis label

image


Custom direction doesn’t seem to work. Angles are greyed out and selecting geometry doesn’t do anything.

**Comment by greggjaskiewicz** [comment]
There’s a big update to this PR coming following Max’s comments on Discord . Standby :-) On 15 Aug 2026, at 10:55, Thomas D. ***@***.***> wrote:﻿pierreporte left a comment (FreeCAD/FreeCAD#28647)
Just compiled. I have the following issues:
Using the assembly example file, solids created at the root of the document have the filling but not the assembly container.
Image: image (view on web)
Hatches seem to be slightly above the cutting plane.
Image: image (view on web)
There shouldn’t be an error if there is no intersection with geometry. Just let the cutting plane sit in the empty.
The gizmo has an empty Z-axis label
Image: image (view on web)
Custom direction doesn’t seem to work. Angles are greyed out and selecting geometry doesn’t do anything.

**Comment by greggjaskiewicz** [comment]
In case it was missed, I pushed a big update - fixing many shortcomings which were highlighted before . As always, due to the amount of changes - some of the prior comments may no longer apply.

File: src/Gui/Inventor/Draggers/Gizmo.cpp [hunk 17]
```diff
@@ -236,6 +236,23 @@ void LinearGizmo::reverseDir()
     getDraggerContainer()->setPointerDirection(dir * -1);
```
Comment (maxwxyz) [inline src/Gui/Inventor/Draggers/Gizmo.cpp:236]: Tesselation is not ideal, on most contours there are artifacts and close geometry colers are overlapping or even are see through.

Image

Image

https://github.com/user-attachments/assets/dc04f46b-c4b4-47d5-ba5d-c483d53e8fc3

Comment (maxwxyz) [inline src/Gui/Inventor/Draggers/Gizmo.cpp:236]: When editing a saved view to change it the draggers are not updated.
Also the blue line for the dragger is in general placed at random positions (maybe origin)


https://github.com/user-attachments/assets/e3feaf95-6d24-406c-9aff-22b57048f73e

Comment (maxwxyz) [inline src/Gui/Inventor/Draggers/Gizmo.cpp:236]: when editing a saved view show hatching setting is not considered (turned off but is active)

Image

Comment (maxwxyz) [inline src/Gui/Inventor/Draggers/Gizmo.cpp:236]: rotation seems to be still around origin

https://github.com/user-attachments/assets/db03af0f-be29-463b-b2af-95f1f4f981ee

Comment (maxwxyz) [inline src/Gui/Inventor/Draggers/Gizmo.cpp:236]: if you would be able to optionally allow to enter manual direction now and to keep it aligned the the camera view while the task is active (even when you orbit) plus adding a distance (z-offset) for that mode, it could instantly replace the useless clipping tool we have today.

Additionally a "no face" mode to see inside solids or select internal elements now hidden by the closed face.
 
Image

Comment (maxwxyz) [inline src/Gui/Inventor/Draggers/Gizmo.cpp:236]: there is an issue with links it seems. It only renders the face but the entire object is hidden or clipped

https://github.com/user-attachments/assets/296a3d5d-7231-4d26-9261-76717e80fdee

Comment (maxwxyz) [inline src/Gui/Inventor/Draggers/Gizmo.cpp:236]: There is an issue with nested objects and adjusted placement


[Example.zip](https://github.com/user-attachments/files/31526847/Example.zip)

https://github.com/user-attachments/assets/29acfc5f-c9b5-4b27-bb90-877adc214b34

Comment (maxwxyz) [inline src/Gui/Inventor/Draggers/Gizmo.cpp:236]: Mesh objects are not clipped currently

Image

Comment (maxwxyz) [inline src/Gui/Inventor/Draggers/Gizmo.cpp:236]: There is an issue where two non-parallel planes do not clip each other

https://github.com/user-attachments/assets/3e405c2f-a290-49e0-91f9-58d283ae8583

Comment (maxwxyz) [inline src/Gui/Inventor/Draggers/Gizmo.cpp:236]: Improvements:

- When I select a face, edge whatever and start the command: align the plane to that plane, edge... (similar as Std_AlignToSelection). Maybe not just align but set placement to this. then offset allows to slide along. this way you can quickly position the plane normal to a cylindrical face (axis) or through two axes...
- when I have result mode "Geometry" there is a different hatching (no big deal though)
- the display mode is somehow not shaded. The geometry result type also considers reflections and stuff. display mode is darker

**Comment by pierreporte** [comment]
When the section plane is on a surface, then we can see the inside of the part.

image

image

**Comment by pierreporte** [comment]
There needs to be a way to align the section plane with an arbitrary reference. I propose again a similar mechanism to the assembly JCS. Currently you need to align the view first with `Std_AlignToSelection` then select view direction. Not practical at all.

Do not display the blue line, it doesn’t make sense IMO.

By default, the section plane can be placed in a position and orientation where nothing is visible. I would recommend to make for a selection of the user first. With the JCS style positioning, then it directly wait for the user to position the gizmo.

**Comment by pierreporte** [comment]
Removed material opacity should be increased significantly. If you activate this option, you want to see it so it must be clearly visible. I don’t know the right value but we could start at 50 % first. Maybe provide a slider? I can see user willing to adjust the value depending on what they do.

**Comment by pierreporte** [comment]
Hatches are made with lines that have a thickness expressed in pixels so if you zoom out, they eventually cover the filling. I propose to make the distance between them also constant regardless of zoom level, and even independent to orientation like Fusion 360.

**Comment by codeotaku** [comment]
Can we not just use vectors instead of pixels for this? This is a cool feature, but blurry pixels/different dpi monitors, etc...

**Comment by greggjaskiewicz** [comment]
loads of v2.0 feature requests, that I'd like to file under "future improvements" indeed.

File: src/Gui/Inventor/Draggers/Gizmo.cpp (hunk 17 above)
Comment (greggjaskiewicz) [reply src/Gui/Inventor/Draggers/Gizmo.cpp:236]: can't reproduce anymore @maxwxyz , please retest after latest change

Comment (maxwxyz) [reply src/Gui/Inventor/Draggers/Gizmo.cpp:236]: @greggjaskiewicz this issue still persists.
1. start section
2. uncheck show hatching
3. click OK to save
4. double click the tree object --> hatching is unchecked but still shown in the 3D view

File: src/Gui/Inventor/Draggers/Gizmo.cpp
Comment (maxwxyz) [inline src/Gui/Inventor/Draggers/Gizmo.cpp:233]: When the clipping plane is rotated and due to the model the size changes, the arrow becomes tiny and not able to be clicked. Zooming out increases the dragger then. Zooming in makes the dragger smaller and smaller:

https://github.com/user-attachments/assets/c5ff042e-a1f5-457b-9c99-7ef47327ff3f

Comment (maxwxyz) [inline src/Gui/Inventor/Draggers/Gizmo.cpp:233]: It would be great to have drag support in local X and Y to be able to rotate around another axis.
Either with linear draggers, a planar face (like transform) or the full blown transform dragger.
Can be fullow up IMO

Image

Image

Comment (maxwxyz) [inline src/Gui/Inventor/Draggers/Gizmo.cpp:230]: when rotating along one axis and then rotating the other axis it is not rotating around the correct axis and the dragger is placed at the wrong angle and rotates differently:

https://github.com/user-attachments/assets/4cfe18df-64d8-4cb5-954f-1ca207ab0984

File: src/Gui/Inventor/Draggers/Gizmo.cpp (hunk 17 above)
Comment (maxwxyz) [reply src/Gui/Inventor/Draggers/Gizmo.cpp:236]: still present

Comment (maxwxyz) [reply src/Gui/Inventor/Draggers/Gizmo.cpp:236]: this is fixed!

**Comment by greggjaskiewicz** [comment]
Tools.cpp change is only to fix temporary crash in main.

File: src/Gui/Inventor/Draggers/Gizmo.cpp
Comment (maxwxyz) [reply src/Gui/Inventor/Draggers/Gizmo.cpp:233]: Fixed

Comment (maxwxyz) [reply src/Gui/Inventor/Draggers/Gizmo.cpp:233]: Dragger works great. As follow up maybe: dragging the plane snaps back instantly:

https://github.com/user-attachments/assets/208d3a76-3aef-4717-962c-6debc3fcdf32

Comment (maxwxyz) [reply src/Gui/Inventor/Draggers/Gizmo.cpp:230]: Fixed

Comment (maxwxyz) [inline src/Gui/Inventor/Draggers/Gizmo.cpp:227]: Not all objects have closed faces after the last push. The same models worked before your push.
They do show their colored outline but no hatches or closed face, I can see inside of them. When setting result mode to geometry it is closed.

Also on objects which are closed, the faces are again not fully closed and you can see inside (similar issue as a few iterations before):


 https://github.com/user-attachments/assets/21e01bf3-7951-472b-8745-56b6c2226ef
